### PR TITLE
Add subscriptions for creating cha+opp

### PR DESCRIPTION
**Pull request by alstojanovic**
### Describe the background of your pull request

What does your pull request do? Does it solve a bug (which one?), add a feature?

### Additional context

Add any other context

### Governance 

- [ ] Documentation is added
- [ ] Test cases are added or updated

By submitting this pull request I confirm that:
- I've read the contributing document https://github.com/alkem-io/.github/blob/master/CONTRIBUTING.md.
- I've read and understand the Code of Conduct https://github.com/alkem-io/.github/blob/master/CODE_OF_CONDUCT.md.
- I understand that any contributions or suggestions I made may make it into the actual code. I've read the License 
     https://github.com/alkem-io/Coordination/blob/master/LICENSE.
- I understand that submitting the pull request requires agreeing to and signing the contributor license agreement.

### CODE REVIEW

File: src/common/constants/providers.ts [hunk 1]
```diff
@@ -14,6 +14,10 @@ export const SUBSCRIPTION_PROFILE_VERIFIED_CREDENTIAL =
   'alkemio-subscriptions-profile-verified-credential';
 export const SUBSCRIPTION_CALLOUT_MESSAGE_CREATED =
   'alkemio-subscriptions-callout-message-created';
+export const SUBSCRIPTION_CHALLENGE_OPPORTUNITY_CREATED =
+  'alkemio-subscriptions-challenge-opportunity-created';
+export const SUBSCRIPTION_HUB_CHALLENGE_CREATED =
+  'alkemio-subscriptions-hub-challenge-created';
 export const NOTIFICATIONS_SERVICE = 'alkemio-notifications';
 export const WALLET_MANAGEMENT_SERVICE = 'alkemio-wallet-manager';
 export const REDIS_LOCK_SERVICE = 'redis-lock-service';
```
Comment (hero101) [inline src/common/constants/providers.ts:23]: any reason to include "challenge" and "hub" in the names?

File: src/common/enums/messaging.queue.ts [hunk 2]
```diff
@@ -9,4 +9,6 @@ export enum MessagingQueue {
   SUBSCRIPTION_PROFILE_VERIFIED_CREDENTIAL = 'alkemio-subscriptions-profile-verified-credential',
   SUBSCRIPTION_CALLOUT_ASPECT_CREATED = 'alkemio-subscriptions-callout-aspect-created',
   SUBSCRIPTION_CALLOUT_MESSAGE_CREATED = 'alkemio-subscriptions-callout-message-created',
+  SUBSCRIPTION_CHALLENGE_OPPORTUNITY_CREATED = 'alkemio-subscriptions-challenge-opportunity-created',
+  SUBSCRIPTION_HUB_CHALLENGE_CREATED = 'alkemio-subscriptions-hub-challenge-created',
```
Comment (hero101) [inline src/common/enums/messaging.queue.ts]: any reason to include "challenge" and "hub" in the names?

File: src/core/microservices/microservices.module.ts [hunk 3]
```diff
@@ -94,6 +104,8 @@ const subscriptionFactoryProviders = subscriptionConfig.map(
     NOTIFICATIONS_SERVICE,
     WALLET_MANAGEMENT_SERVICE,
     NOTIFICATIONS_SERVICE,
+    SUBSCRIPTION_CHALLENGE_OPPORTUNITY_CREATED,
+    SUBSCRIPTION_HUB_CHALLENGE_CREATED,
```
Comment (hero101) [inline src/core/microservices/microservices.module.ts]: move them next to the other subs

File: src/domain/challenge/challenge/challenge.resolver.mutations.ts [hunk 4]
```diff
@@ -91,10 +97,24 @@ export class ChallengeResolverMutations {
...
     );
+
+    const opportunityCreatedEvent: OpportunityCreatedPayload = {
+      eventID: `challenge-opportunity-created-${Math.round(
+        Math.random() * 100
+      )}`,
+      challengeID: challenge.id,
+      opportunity,
+    };
+    await this.opportunityCreatedSubscription.publish(
```
Comment (hero101) [inline src/domain/challenge/challenge/challenge.resolver.mutations.ts]: i don't think you need to wait for this

File: src/domain/challenge/challenge/challenge.resolver.mutations.ts [hunk 5]
```diff
@@ -91,10 +97,24 @@ export class ChallengeResolverMutations {
...
       agentInfo
     );
-    return await this.opportunityAuthorizationService.applyAuthorizationPolicy(
+    await this.opportunityAuthorizationService.applyAuthorizationPolicy(
       opportunity,
       challenge.authorization
     );
+
+    const opportunityCreatedEvent: OpportunityCreatedPayload = {
+      eventID: `challenge-opportunity-created-${Math.round(
```
Comment (hero101) [inline src/domain/challenge/challenge/challenge.resolver.mutations.ts]: is 'challenge' needed in the name?

File: src/domain/challenge/challenge/challenge.resolver.subscriptions.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    private challengeService: ChallengeService,
+    private authorizationService: AuthorizationService
+  ) {}
+
+  @UseGuards(GraphqlGuard)
+  @TypedSubscription<OpportunityCreatedPayload, OpportunityCreatedArgs>(
+    () => OpportunityCreated,
+    {
+      description:
+        'Receive new Update messages on Communities the currently authenticated User is a member of.',
```
Comment (hero101) [inline src/domain/challenge/challenge/challenge.resolver.subscriptions.ts]: copy pasted

File: src/domain/challenge/challenge/challenge.resolver.subscriptions.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+          `${logMsgPrefix} Filter result is ${isSameChallenge}`,
+          LogContext.SUBSCRIPTIONS
+        );
+        return isSameChallenge;
+      },
+    }
+  )
+  async opportunityCreated(
+    @CurrentUser() agentInfo: AgentInfo,
+    @Args({
```
Comment (hero101) [inline src/domain/challenge/challenge/challenge.resolver.subscriptions.ts:68]: `OpportunityCreatedArgs`

File: src/domain/challenge/challenge/challenge.resolver.subscriptions.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+  @TypedSubscription<OpportunityCreatedPayload, OpportunityCreatedArgs>(
+    () => OpportunityCreated,
+    {
+      description:
+        'Receive new Update messages on Communities the currently authenticated User is a member of.',
+      resolve(this: ChallengeResolverSubscriptions, payload, args, context) {
+        const agentInfo = context.req.user;
+        const logMsgPrefix = `[User (${agentInfo.email}) Challenge Opportunities] - `;
+        this.logger.verbose?.(
+          `${logMsgPrefix} sending out event for Opportunities on Challenge: ${payload.challengeID} `,
```
Comment (hero101) [inline src/domain/challenge/challenge/challenge.resolver.subscriptions.ts]: this message can be improved

File: src/domain/challenge/challenge/challenge.resolver.subscriptions.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+  @UseGuards(GraphqlGuard)
+  @TypedSubscription<OpportunityCreatedPayload, OpportunityCreatedArgs>(
+    () => OpportunityCreated,
+    {
+      description:
+        'Receive new Update messages on Communities the currently authenticated User is a member of.',
+      resolve(this: ChallengeResolverSubscriptions, payload, args, context) {
+        const agentInfo = context.req.user;
+        const logMsgPrefix = `[User (${agentInfo.email}) Challenge Opportunities] - `;
```
Comment (hero101) [inline src/domain/challenge/challenge/challenge.resolver.subscriptions.ts]: this message can be improved

File: src/domain/challenge/challenge/challenge.resolver.subscriptions.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        return payload;
+      },
+      filter(
+        this: ChallengeResolverSubscriptions,
+        payload,
+        variables,
+        context
+      ) {
+        const agentInfo = context.req.user;
+        const logMsgPrefix = `[User (${agentInfo.email}) Challenge Opportunities] - `;
```
Comment (hero101) [inline src/domain/challenge/challenge/challenge.resolver.subscriptions.ts]: this message can be improved

File: src/domain/challenge/challenge/challenge.resolver.subscriptions.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    @CurrentUser() agentInfo: AgentInfo,
+    @Args({
+      name: 'challengeID',
+      type: () => UUID,
+      description: 'The ID of the Challenge to subscribe to.',
+      nullable: false,
+    })
+    challengeID: string
+  ) {
+    const logMsgPrefix = `[User (${agentInfo.email}) Challenge Opportunities] - `;
```
Comment (hero101) [inline src/domain/challenge/challenge/challenge.resolver.subscriptions.ts]: this message can be improved

File: src/domain/challenge/hub/hub.resolver.mutations.ts [hunk 12]
```diff
@@ -242,11 +248,24 @@ export class HubResolverMutations {
...
       hub.authorization,
       hubCommunityCredential
     );
+
+    const challengeCreatedEvent: ChallengeCreatedPayload = {
+      eventID: `hub-challenge-created-${Math.round(Math.random() * 100)}`,
+      hubID: hub.id,
+      challenge,
+    };
+    await this.challengeCreatedSubscription.publish(
```
Comment (hero101) [inline src/domain/challenge/hub/hub.resolver.mutations.ts]: i don't think this needs to be awaited

File: src/domain/challenge/hub/hub.resolver.subscriptions.ts [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    private hubService: HubService,
+    private authorizationService: AuthorizationService
+  ) {}
+
+  @UseGuards(GraphqlGuard)
+  @TypedSubscription<ChallengeCreatedPayload, ChallengeCreatedArgs>(
+    () => ChallengeCreated,
+    {
+      description:
+        'Receive new Update messages on Communities the currently authenticated User is a member of.',
```
Comment (hero101) [inline src/domain/challenge/hub/hub.resolver.subscriptions.ts]: copy pasted

File: src/domain/challenge/hub/hub.resolver.subscriptions.ts [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+      resolve(this: HubResolverSubscriptions, payload, args, context) {
+        const agentInfo = context.req.user;
+        const logMsgPrefix = `[User (${agentInfo.email}) Hub Challenges] - `;
+        this.logger.verbose?.(
+          `${logMsgPrefix} sending out event for Challenges on Hub: ${payload.hubID} `,
```
Comment (hero101) [inline src/domain/challenge/hub/hub.resolver.subscriptions.ts]: this message can be improved

File: src/domain/challenge/hub/hub.resolver.subscriptions.ts [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+  @UseGuards(GraphqlGuard)
+  @TypedSubscription<ChallengeCreatedPayload, ChallengeCreatedArgs>(
+    () => ChallengeCreated,
+    {
+      description:
+        'Receive new Update messages on Communities the currently authenticated User is a member of.',
+      resolve(this: HubResolverSubscriptions, payload, args, context) {
+        const agentInfo = context.req.user;
+        const logMsgPrefix = `[User (${agentInfo.email}) Hub Challenges] - `;
```
Comment (hero101) [inline src/domain/challenge/hub/hub.resolver.subscriptions.ts]: this message can be improved

File: src/domain/challenge/hub/hub.resolver.subscriptions.ts [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+        this.logger.verbose?.(
+          `${logMsgPrefix} sending out event for Challenges on Hub: ${payload.hubID} `,
+          LogContext.SUBSCRIPTIONS
+        );
+        return payload;
+      },
+      filter(this: HubResolverSubscriptions, payload, variables, context) {
+        const agentInfo = context.req.user;
+        const logMsgPrefix = `[User (${agentInfo.email}) Hub Challenges] - `;
```
Comment (hero101) [inline src/domain/challenge/hub/hub.resolver.subscriptions.ts]: this message can be improved

File: src/domain/challenge/hub/hub.resolver.subscriptions.ts [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+          `${logMsgPrefix} Filter result is ${isSameHub}`,
+          LogContext.SUBSCRIPTIONS
+        );
+        return isSameHub;
+      },
+    }
+  )
+  async challengeCreated(
+    @CurrentUser() agentInfo: AgentInfo,
+    @Args({
```
Comment (hero101) [inline src/domain/challenge/hub/hub.resolver.subscriptions.ts:63]: `ChallengeCreatedArgs`

File: src/domain/challenge/hub/hub.resolver.subscriptions.ts [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+      name: 'hubID',
+      type: () => UUID,
+      description: 'The ID of the Hub to subscribe to.',
+      nullable: false,
+    })
+    hubID: string
+  ) {
+    const logMsgPrefix = `[User (${agentInfo.email}) Hub Challenges] - `;
```
Comment (hero101) [inline src/domain/challenge/hub/hub.resolver.subscriptions.ts]: this message can be improved

File: src/domain/challenge/hub/hub.resolver.subscriptions.ts [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+    this.logger.verbose?.(
+      `${logMsgPrefix} Subscribing to Challenges on the following Hub: ${hubID}`,
```
Comment (hero101) [inline src/domain/challenge/hub/hub.resolver.subscriptions.ts]: this message can be improved

File: src/domain/challenge/hub/dto/hub.dto.event.challenge.created.ts [hunk 20]
```diff
@@ -0,0 +1,20 @@
+import { Field, ObjectType } from '@nestjs/graphql';
+import { IChallenge } from '@domain/challenge/challenge/challenge.interface';
+
+@ObjectType('ChallengeCreated')
+export class ChallengeCreated {
+  eventID!: string;
+
+  @Field(() => String, {
```
Comment (hero101) [inline src/domain/challenge/hub/dto/hub.dto.event.challenge.created.ts]: `UUID_NAMEID`

File: src/domain/challenge/hub/dto/hub.challenge.created.args.ts [hunk 21]
```diff
@@ -0,0 +1,11 @@
+import { ArgsType, Field } from '@nestjs/graphql';
+import { UUID } from '@domain/common/scalars';
+
+@ArgsType()
+export class ChallengeCreatedArgs {
+  @Field(() => UUID, {
```
Comment (hero101) [inline src/domain/challenge/hub/dto/hub.challenge.created.args.ts]: `UUID_NAMEID`

File: src/domain/challenge/challenge/dto/challenge.dto.event.opportunity.created.ts [hunk 22]
```diff
@@ -0,0 +1,20 @@
+import { Field, ObjectType } from '@nestjs/graphql';
+import { IOpportunity } from '@domain/collaboration/opportunity/opportunity.interface';
+
+@ObjectType('OpportunityCreated')
+export class OpportunityCreated {
+  eventID!: string;
+
+  @Field(() => String, {
```
Comment (hero101) [inline src/domain/challenge/challenge/dto/challenge.dto.event.opportunity.created.ts]: `UUID`

**Review by hero101 — CHANGES_REQUESTED** [review]
I have tested the code locally.

File: src/common/constants/providers.ts (hunk 1 above)
Comment (alstojanovic) [reply src/common/constants/providers.ts:23]: No reason, removing the prefixes.

File: src/common/enums/messaging.queue.ts (hunk 2 above)
Comment (alstojanovic) [reply src/common/enums/messaging.queue.ts]: removing them.

File: src/domain/challenge/challenge/challenge.resolver.mutations.ts (hunk 4 above)
Comment (alstojanovic) [reply src/domain/challenge/challenge/challenge.resolver.mutations.ts]: Well, no I don't. I don't depend on whether the created challenge is published.

File: src/domain/challenge/challenge/dto/challenge.dto.event.opportunity.created.ts [hunk 23]
```diff
@@ -0,0 +1,21 @@
+import { Field, ObjectType } from '@nestjs/graphql';
+import { IOpportunity } from '@domain/collaboration/opportunity/opportunity.interface';
+import { UUID_NAMEID } from '@domain/common/scalars/scalar.uuid.nameid';
+
+@ObjectType('OpportunityCreated')
+export class OpportunityCreated {
+  eventID!: string;
+
+  @Field(() => UUID_NAMEID, {
```
Comment (hero101) [inline src/domain/challenge/challenge/dto/challenge.dto.event.opportunity.created.ts]: the reason I suggested `UUID` and not `UUID_NAMEID` was that the nameID is unique ONLY in the context of a hub, so you can have multiple challenges with the same nameID.

File: src/domain/challenge/challenge/dto/challenge.opportunity.created.args.ts [hunk 24]
```diff
@@ -0,0 +1,11 @@
+import { ArgsType, Field } from '@nestjs/graphql';
+import { UUID_NAMEID } from '@domain/common/scalars';
+
+@ArgsType()
+export class OpportunityCreatedArgs {
+  @Field(() => UUID_NAMEID, {
```
Comment (hero101) [inline src/domain/challenge/challenge/dto/challenge.opportunity.created.args.ts]: the reason I suggested `UUID` and not `UUID_NAMEID` was that the nameID is unique ONLY in the context of a hub, so you can have multiple challenges with the same nameID.

**Review by hero101 — CHANGES_REQUESTED** [review]

**Review by hero101 — DISMISSED** [review]

**Review by hero101 — APPROVED** [review]

**Pull request merged**